controllers/hcloudremediation_controller_test.go

```diff
@@ -0,0 +1,268 @@
+			Expect(testEnv.Create(ctx, hcloudRemediation)).To(Succeed())
+
+			Eventually(func() bool {
+				if err := testEnv.Get(ctx, hcloudRemediationkey, hcloudRemediation); err != nil {
+					return false
+				}
+
+				return true
+			}, timeout).Should(BeTrue())
+		})
+
+		It("checks if hcloudRemediation objects has the HCloudTokenAvailableCondition condition", func() {
+			Expect(testEnv.Create(ctx, hcloudRemediation)).To(Succeed())
+
+			Eventually(func() bool {
+				if err := testEnv.Get(ctx, hcloudRemediationkey, hcloudRemediation); err != nil {
+					return false
+				}
+
+				testEnv.GetLogger().Info("Status condition of the hcloud remediation", "condition", hcloudRemediation.Status.Conditions)
+				return isPresentAndTrue(hcloudRemediationkey, hcloudRemediation, infrav1.HCloudTokenAvailableCondition)
+			}, timeout).Should(BeTrue())
+		})
+
+		It("checks for deleting phase of hcloudRemediation object and MachineOwnerRemediatedCondition of capiMachine object", func() {
+			Eventually(func() bool {
+				if err := testEnv.Get(ctx, hcloudMachineKey, hcloudMachine); err != nil {
+					return false
+				}
+
+				testEnv.GetLogger().Info("Status of the hcloudmachine", "status", hcloudMachine.Status)
+				return hcloudMachine.Status.Ready
+			}, timeout).Should(BeTrue())
+
+			pID, err := hcloudutil.ServerIDFromProviderID(hcloudMachine.Spec.ProviderID)
+			Expect(err).NotTo(HaveOccurred())
+
+			Expect(hcloudClient.DeleteServer(ctx, &hcloud.Server{ID: pID})).NotTo(HaveOccurred())
+
+			Expect(testEnv.Create(ctx, hcloudRemediation)).To(Succeed())
+
+			Eventually(func() bool {
+				if err := testEnv.Get(ctx, hcloudRemediationkey, hcloudRemediation); err != nil {
+					return false
+				}
+
+				return hcloudRemediation.Status.Phase == infrav1.PhaseDeleting && isPresentAndFalseWithReason(capiMachineKey, capiMachine, clusterv1.MachineOwnerRemediatedCondition, clusterv1.WaitingForRemediationReason)
+			}, timeout).Should(BeTrue())
+		})
+
+		It("check if retryCount is one and LastRemediated is set", func() {
+			Expect(testEnv.Create(ctx, hcloudRemediation)).To(Succeed())
+
+			Eventually(func() bool {
+				if err := testEnv.Get(ctx, hcloudRemediationkey, hcloudRemediation); err != nil {
+					return false
+				}
+
+				return hcloudRemediation.Status.LastRemediated != nil && hcloudRemediation.Status.RetryCount == 1
+			}, timeout).Should(BeTrue())
+		})
+
+		It("check if PhaseWaiting is set when retryLimit is reached", func() {
+			Expect(testEnv.Create(ctx, hcloudRemediation)).To(Succeed())
+
+			Eventually(func() bool {
+				if err := testEnv.Get(ctx, hcloudRemediationkey, hcloudRemediation); err != nil {
+					return false
+				}
+
+				testEnv.GetLogger().Info("status of hcloudRemediation", "status", hcloudRemediation.Status.Phase)
+				return hcloudRemediation.Status.Phase == infrav1.PhaseWaiting
+			}, timeout).Should(BeTrue())
+		})
+
+		It("check if MachineOwnerRemediatedCondition is set on capiMachine for phase waiting", func() {
```

Comment:
I don't get how this works: You have the same test here and above, in both cases you have PhaseWaiting. However, in this test you get immediately from waiting to deleting?! 
I looked into the code and the actual difference is that you look at the LastRemediated timepoint and decide based on that whether you go from "PhaseWaiting" in "PhaseDeleting", right? 
this should be definitely incorporated in your test somehow.

Reply:
https://github.com/syself/cluster-api-provider-hetzner/blob/4e2e93e453430610cb065ad54926204cb515159d/pkg/services/hcloud/remediation/remediation.go#L143 I used this as a reference, the test above also sets the phase == waiting, so, if I add another eventually that checks the logic after setting to waiting phase? - 


```go
FIt("checks if PhaseWaiting is set when retryLimit is reached", func() {
			Expect(testEnv.Create(ctx, hcloudRemediation)).To(Succeed())

			Eventually(func() bool {
				if err := testEnv.Get(ctx, hcloudRemediationkey, hcloudRemediation); err != nil {
					return false
				}

				testEnv.GetLogger().Info("status of hcloudRemediation", "status", hcloudRemediation.Status.Phase)
				return hcloudRemediation.Status.Phase == infrav1.PhaseWaiting
			}, timeout).Should(BeTrue())

			Eventually(func() bool {
				if err := testEnv.Get(ctx, hcloudRemediationkey, hcloudRemediation); err != nil {
					return false
				}

				testEnv.GetLogger().Info("status of hcloudRemediation", "status", hcloudRemediation.Status.Phase)
				return hcloudRemediation.Status.Phase == infrav1.PhaseDeleting && isPresentAndFalseWithReason(capiMachineKey, capiMachine, clusterv1.MachineOwnerRemediatedCondition, clusterv1.WaitingForRemediationReason)
			}, timeout).Should(BeTrue())
		})
```

Reply:
no it should be two different tests.. but you should just include the fact that what you do here only works because of this LastRemediated.. Otherwise it wouldn't be "waiting" but "deleting".. and it is only "deleting" and not "waiting" because the LastRemediated time is "long ago"